### ISSUE TITLE
[gap] P0: every strategy silently uses wrong-interval indicators whenever multiple candle_intervals close for the same symbol (default config)

### ISSUE THREAD

**Issue by SakethThogarucheeti** [SakethThogarucheeti/trading-platform#36]
## What

`Settings.candle_intervals` defaults to `["1min", "5min", "15min"]`
(`src/trading/config/settings.py:88`), and there is exactly **one** shared
`CandleAggregator` per running system, constructed with this full interval
list (`di/containers/components.py:86`/`217`). Per-algo `AlgoSettings.candle_intervals`
(`config/settings.py:54`) looks like it scopes which intervals that specific
algo trades, but **it does not** — it's only ever used as descriptive metadata
persisted to `algo_configs.candle_intervals` (`di/providers/algo_pipeline.py:151`,
via `seed_state`) and for warmup lookback sizing
(`candles/service/aggregator.py:107`, `_fetch_warmup_candles`). It is **never**
used to filter which candle-close events actually reach that algo's strategy.

Confirmed the live dispatch path end to end:

```python
# trading/candles/service/aggregator.py — CandleAggregator.handle()
async def handle(self, tick: TickEvent) -> list[CandleEvent]:
    ...
    closed: list[CandleEvent] = []
    for interval in self._config.intervals:        # ALL global intervals, every tick
        candle = self._accumulator.process(sc, interval, tick)
        if candle is not None:
            closed.append(candle)
    return closed

# trading/app/pipeline.py — TickPipeline.run()
async def run(self, tick: TickEvent) -> None:
    candles = await self._candle_registry.handle(tick)
    for candle in candles:                          # every closed interval, no filter
        signals = await self._signal_generator.handle(candle)
        await self._algo_pipeline.run(signals)
```

`tst/unit/core/test_pipeline.py::test_run_processes_all_candles_when_multiple_intervals_close`
confirms this fan-out IS the deliberately tested, intended dispatch behavior — every
algo's `SignalGenerator` really does receive a `1min` **and** a `5min` **and** a `15min`
`CandleEvent` for the same symbol, back to back, under the default config. That test
mocks `SignalGenerator` entirely, though, so it only verifies the dispatch plumbing —
it does not exercise (and therefore cannot catch) what happens once a real strategy
receives multiple intervals for the same symbol.

## The bug: two places silently conflate state across intervals

**1. `trading_strategy_sdk/indicator_cache.py` — `IndicatorCacheMixin._get_inds`**
(used by all 7 strategies via `on_candle(symbol, instrument_type, candle)` →
`self._get_inds(symbol, candle.interval)`):

```python
def _get_inds(self, symbol: str, interval: str) -> IndTuple:
    if symbol not in self._inds:                     # keyed by symbol ONLY
        assert self._store is not None, ...
        self._inds[symbol] = self._build_indicators(self._store, symbol, interval)
    return self._inds[symbol]
```

Indicator objects are constructed bound to a specific `interval`
(e.g. `EMA(store, symbol, interval)`, `ATR(store, symbol, interval)` —
every strategy's `_build_indicators` does this). The cache key is `symbol`
alone. So: the first candle to close for a symbol (whichever interval wins
the race — order isn't guaranteed) builds and caches indicators bound to
that interval. Every subsequent candle-close for that same symbol at a
*different* interval hits the cache and silently reuses indicators still
internally bound to the *first* interval — computing signal decisions and
stop-distances (`atr * atr_multiplier`) from the wrong timeframe's data,
while `candle.close`/`candle.timestamp` (used for `entry_price`/logging)
correctly reflect the *actual* interval. The mismatch is invisible: no
exception, no log line, just wrong numbers silently feeding into live
order sizing.

**2. `trading_strategy_sdk/base.py` — `AlgoInstance.tick_bar`**:

```python
def tick_bar(self, interval: str, warmup_candles: int) -> None:
    self.interval = interval                          # last interval wins
    self.bars_seen += 1                                # shared across ALL intervals
    if self.bars_seen >= warmup_candles:
        self.warmed_up = True
```

`bars_seen`/`warmed_up` are single counters shared across every interval's
candle closes for that symbol — so "warmup complete" fires after
`warmup_candles` *combined* 1min+5min+15min closes, not after
`warmup_candles` bars of any one meaningful timeframe. `self.interval`
(singular) just reflects whichever interval closed most recently, and is
later used by `restore_state()`
(`trading/strategy/service/generator.py:229-233`) to look up the rolling-
state cache under a single, non-deterministic interval key — inconsistent
with whatever was actually saved for the *other* intervals.

## Reachability

This isn't a rare multi-interval-config edge case — `Settings.candle_intervals`
defaults to 3 intervals, and per-algo `candle_intervals` doesn't actually
scope anything (see above), so **every algo, under default configuration,
hits this on every symbol it trades**, for at least 2 of every 3 candle
closes.

## Why this needs a human call, not a mechanical loop fix

The unambiguous part is the bug itself — caching interval-bound state by
symbol alone is wrong regardless of intent. What's NOT clear from the code
is the intended architecture, and the correct fix depends entirely on it:

1. **Strategies are meant to be genuinely multi-interval-aware** (a
   1min-EMA-crossover-that-also-watches-15min-trend kind of design) — fix:
   key `IndicatorCacheMixin._get_inds` by `(symbol, interval)`, and give
   `AlgoInstance` per-interval `bars_seen`/`warmed_up`/`interval` state
   (a dict keyed by interval, not scalars) throughout
   `tick_bar`/`state_dict`/`restore_state`. Larger change, touches the
   rolling-state persistence format too.
2. **Each algo is meant to trade exactly ONE interval**, and
   `algo.candle_intervals` was *supposed* to filter incoming candles but
   the filter was never wired up — fix: add an interval check in
   `SignalGenerator.handle()` (or `_tick_bar_and_check_ready`) that drops
   candles whose `interval` isn't in the algo's configured set, before
   they ever reach `on_candle()`/`tick_bar()`. Much smaller change, but
   changes what "warmup complete" / live trading actually does today for
   every running algo — needs sign-off given the live money impact.

Either direction is a live-trading-behavior change, not a local refactor —
flagging for a human decision rather than guessing given the blast radius
(every algo, every symbol, indicator correctness and stop-distance sizing
on real orders).

## Suggested immediate mitigation (lower-risk than the full fix, still needs sign-off)

If a quick stopgap is wanted before the architectural decision is made:
setting every algo's `candle_intervals` explicitly to a single-entry list
matching `settings.candle_intervals[0]` ("1min") wouldn't fix the
underlying filter gap (candles for 5min/15min still get dispatched and
still corrupt the shared cache/counters) — it does NOT mitigate this on
its own. There isn't a safe config-only workaround; this needs a code fix.

**Comment by SakethThogarucheeti** [comment]
## Investigate-mode follow-up (round 134, 2026-09-12)

Re-verified the whole chain against current `main` — nothing has drifted since this issue
was filed (2026-08-24); `_get_inds`, `AlgoInstance.tick_bar`, `_tick_bar_and_check_ready` are
all byte-identical to what's quoted above. Two new grounded findings while tracing the fix
options, plus a granular breakdown and top-3 design options as requested.

### New findings (not in the original writeup)

1. **`AlgoPipelineFactory.build_pipeline`'s own `intervals: list[str]` parameter is dead.**
   Read the full body (`di/providers/algo_pipeline.py:58-134`) — `intervals` is accepted but
   never referenced anywhere inside the method. It's only used by the *other* branch of the
   caller, `seed_state()` (metadata persisted to `algo_configs.candle_intervals`) and by
   `CandleAggregatorComponent`'s warmup sizing. This confirms, at the parameter level, that
   per-algo interval scoping was wired as far as "pass it into the constructor call" and no
   further — `SignalGenerator` never receives it at all.

2. **`SignalGenerator.restore_state()` is never called anywhere in production code.**
   `grep -rn restore_state src/` across all of trading-platform turns up exactly 3 hits: the
   method definition itself, a docstring in `storage/cache/rolling_state.py` ("Consumers
   (restore_state at startup) call load_latest()..."), and a claim in
   `strategy/README.md` ("Strategy state ... can be restored across restarts via
   `SignalGenerator.restore_state()`"). No call site exists in `build_runtime`,
   `_RuntimeAssembler`, or any `Component._setup()` — and no test calls it either. Today, a
   restart discards all rolling strategy state regardless of the interval-keying bug; the
   `instance.interval`-keyed lookup this issue flagged as "inconsistent" is actually
   unreachable in practice. This is a distinct, mechanically-fixable bug (wire up one call at
   the right point in startup) rather than a 3-way design question on its own — but it directly
   bears on options below, since whichever interval-state shape gets chosen also needs its
   restore path actually invoked to mean anything. Recommend filing this as its own follow-up
   issue once an option here is picked, since the correct call site depends on the chosen
   state shape.

### Granular sub-problems

1. **Intent** — is a "strategy" meant to trade exactly one interval (today's only actually-used
   shape — every current strategy's `_build_indicators` takes one `interval` argument, per
   `indicator_cache.py`), or genuinely watch multiple intervals for one symbol at once (e.g. a
   1min entry confirmed by a 15min trend filter)? Nothing in the SDK's 7 concrete strategies
   does the latter today.
2. **Dispatch filtering** — should `SignalGenerator`/`_tick_bar_and_check_ready` drop candles
   for intervals an algo isn't configured for, before they ever reach `tick_bar()`/`on_candle()`?
   Today nothing filters — every candle for a matched symbol reaches every algo trading that
   symbol, regardless of interval.
3. **Cache/state keying** — `IndicatorCacheMixin._get_inds` (keyed by `symbol` only) and
   `AlgoInstance.bars_seen`/`warmed_up`/`interval` (scalar, not per-interval) both assume single-
   interval dispatch. If sub-problem 2 stays unfiltered, these need to become interval-aware
   (dict-keyed). If sub-problem 2 gets filtered to one interval per algo, these can stay scalar
   as-is — they become correct by construction since only one interval's candles would ever
   reach them.
4. **Config validation** — `AlgoSettings.candle_intervals: list[str] | None` currently allows
   (and defaults to, via `None` → global list) a *multi*-interval resolution per algo. Whichever
   option is chosen, an algo with no explicit override needs to land on a single interval
   deterministically (not "whichever candle closes first wins the cache," today's actual
   behavior) — this needs either a validator rejecting multi-entry lists, or a design that
   genuinely handles multi-entry correctly.
5. **restore_state() wiring** — see finding #2 above; a real but small companion fix, not itself
   a 3-way design choice, but its correct call site/shape depends on which option below is picked.

### Top 3 design options

**Option 1 — Single-interval-per-algo, filter at dispatch (recommended for now)**
- Add an explicit interval check in `_tick_bar_and_check_ready` (or immediately before it in
  `handle`): if `candle.interval` isn't the algo's one configured interval, return `None`
  without touching `tick_bar()`/the indicator cache at all.
- Change `AlgoSettings.candle_intervals` validation to require exactly one entry when set, and
  make "not set" resolve to `[settings.candle_intervals[0]]` (first global interval) rather
  than the full list — closing sub-problem 4.
- `IndicatorCacheMixin`/`AlgoInstance` need **no structural change** — once only one interval's
  candles ever reach an algo, symbol-only keying and scalar `bars_seen`/`interval` are already
  correct.
- Fix `restore_state()`'s dead-call bug as a companion change: call it once during startup
  after each algo's `SignalGenerator.setup()`, since the interval key is now stable per algo.
- Tradeoffs: smallest diff, no persisted-state format change, immediately closes the live-money
  correctness gap. Forecloses genuine multi-timeframe strategies for now (a strategy wanting a
  second interval would need to `fetch()` it directly from the shared `PolarsStore` out-of-band,
  which already carries all 3 intervals' rows — technically possible today via direct store
  access, just not via the tick-dispatch path).

**Option 2 — Full multi-interval awareness**
- Key `IndicatorCacheMixin._get_inds` by `(symbol, interval)` instead of `symbol` alone.
- Replace `AlgoInstance`'s scalar `interval`/`bars_seen`/`warmed_up` with per-interval dicts;
  update `tick_bar`, `state_dict`, and `restore_state`'s lookup accordingly.
- Requires a migration plan for the persisted `algo_state.state` JSON blob (shape changes from
  scalar fields to a per-interval structure) and the `rolling_state` cache key shape.
- Fix `restore_state()`'s dead-call bug as part of the same change (the state shape is already
  being touched).
- Tradeoffs: correctly supports strategies that genuinely want more than one timeframe per
  symbol — a real future capability — but no current strategy needs it, so this is the larger,
  more speculative option. Bigger blast radius (state format, all 7 strategies' `_build_indicators`
  signatures unaffected but the caching layer around them changes meaningfully), more surface
  for the independent reviewer to check, harder to verify quickly given no strategy exercises
  the multi-interval path today to test against.

**Option 3 — Filter at dispatch, but keep `candle_intervals` genuinely multi-valued per algo**
- Same dispatch-filter mechanism as Option 1, but instead of forcing single-entry lists,
  literally filter candles down to whatever `algo.candle_intervals` (or the resolved global
  default) actually contains — and combine with Option 2's per-interval keying so a
  multi-interval algo config actually works correctly rather than silently degrading.
- This is really "Option 1's filter + Option 2's keying," offered as a middle ground: closes
  the live-correctness gap immediately (no algo silently corrupts cross-interval state) while
  not foreclosing multi-interval configs the way Option 1 does.
- Tradeoffs: same state-migration cost as Option 2, plus the filter logic from Option 1 — more
  moving parts than either pure option, harder to review as one focused change. Only worth it
  if there's a concrete near-term plan to actually write a multi-timeframe strategy; otherwise
  it's paying Option 2's cost without Option 1's simplicity benefit.

### Recommendation

**Option 1**, for the same reason the issue's own "why this needs a human call" section
points at: this is a live-trading behavior change either way, so the smaller, more reviewable
diff should win absent an actual near-term need for multi-timeframe strategies. Every strategy
in the SDK today is single-interval by construction (`_build_indicators(store, symbol,
interval)` — one `interval`), so Option 2/3's added complexity has no current consumer. Pair it
with the `restore_state()` wiring fix as a companion change, and file that piece as its own
tracked issue once this option is confirmed, since its call site depends on this pick.

**Not implemented — proposal only, awaiting human pick**, consistent with every other design
writeup this loop has posted.
